### PR TITLE
Embed webassets natively into teleport instead of attaching to the binary

### LINKED ISSUE #3158: Can't sign Teleport binary for Mac due to malformed Mach-O binary
Signing the `tsh` and `tctl` binaries works fine, but attempts to sign the `teleport` binary on MacOS are failing with the error `main executable failed strict validation`

Apparently this error is raised when the executable being signed does not conform to Apple's strict Mach-O layout rules (https://github.com/zeit/pkg/issues/128) - I highly suspect that this is because of the way we zip the web assets and tack them onto the end of the binary as part of the build process. We had a similar sort of problem when initially building RPMs - when you run `rpmbuild`, the default config tries to strip symbols from the binary. In our case, it ended up stripping the web assets because they just look like junk stuck on the end of the file.

Apple will only notarize a package which has a signed payload, so for now we can't sign a full `teleport` archive (although the client only `tsh` package I'm working on should be fine)

One idea would be to use something like https://github.com/shurcooL/vfsgen to build the assets directly into the binary rather than the current method.

### LINKED ISSUE #5595: Replace custom web assets vendoring with Go's 1.16 Embed
### What

Use  https://golang.org/pkg/embed/ to embed JS assets

### PR DESCRIPTION
**Pull request by a-palchikov**
Use `embed` package to embed webassets instead of ad-hoc attaching to binary.

This PR introduces the following changes:
 * Switch to go1.16 as a prerequisite for the following item.
 * Use embed package to embed webassets instead of ad-hoc attaching to binary

This is required to support building FIPS binaries on mac OS with full code/product sign and notarization. With webassets attached directly to the binary these steps are not possible since the binary format changes and will not be recognized by the tooling/process.

Requires https://github.com/gravitational/teleport.e/pull/231.
Updates #3520, #5595 and #3158.

### CODE REVIEW

**Comment by webvictim** [comment]
As a quick sanity check, this compiles and runs fine for me on`darwin/arm64`.

**Comment by russjones** [comment]
@a-palchikov @webvictim Go 1.16.1 is out but it was a security release, 1.16.2 is supposed to be out within a few days which should be the actual follow-up bug fix release.

I propose we wait a few days (for BoringCrypto release to show up as well) then bump to Go 1.16.2.

What do you two think?

**Comment by russjones** [comment]
@fspmarshall @awly Can you two review?

**Review by fspmarshall — APPROVED** [review]

**Comment by a-palchikov** [comment]
> @a-palchikov @webvictim Go 1.16.1 is out but it was a security release, 1.16.2 is supposed to be out within a few days which should be the actual follow-up bug fix release.
> 
> I propose we wait a few days (for BoringCrypto release to show up as well) then bump to Go 1.16.2.
> 
> What do you two think?

Sure thing.

**Comment by webvictim** [comment]
@russjones Sounds good. It looks like BoringCrypto 1.16 only just came out two days ago, so hopefully it won't be long before the 1.16.1/1.16.2 changes get implemented.

**Comment by russjones** [comment]
@webvictim BoringCrypto 1.16.1 was available when I checked last night, so it should be ready soon after 1.16.2 comes out I think.

File: .gitignore [hunk 1]
```diff
@@ -5,6 +5,9 @@ debug
 default.etcd
 darwin
 
+# web asset directory
+lib/web/build
```
Comment (awly) [inline .gitignore:9]:
```suggestion
/lib/web/build
```

File: build.assets/Dockerfile [hunk 2]
```diff
@@ -97,7 +97,7 @@ ENV GOGOPROTO_ROOT ${GOPATH}/src/github.com/gogo/protobuf
 RUN (curl -L -o /tmp/${PROTOC_TARBALL} https://github.com/google/protobuf/releases/download/v${PROTOC_VER}/${PROTOC_TARBALL} && \
      cd /tmp && unzip /tmp/${PROTOC_TARBALL} -d /usr/local && \
      rm /tmp/${PROTOC_TARBALL})
-RUN (go get -u github.com/gogo/protobuf/proto github.com/gogo/protobuf/protoc-gen-gogo github.com/gogo/protobuf/gogoproto golang.org/x/tools/cmd/goimports && \
+RUN (git clone https://github.com/gogo/protobuf.git ${GOPATH}/src/github.com/gogo/protobuf && go install golang.org/x/tools/cmd/goimports@latest && \
```
Comment (awly) [inline build.assets/Dockerfile:100]:
```suggestion
RUN (git clone --depth=1 https://github.com/gogo/protobuf.git ${GOPATH}/src/github.com/gogo/protobuf && go install golang.org/x/tools/cmd/goimports@latest && \
```

File: lib/web/static.go [hunk 3]
```diff
@@ -12,192 +14,18 @@ distributed under the License is distributed on an "AS IS" BASIS,
...
-	reader, err := f.Open()
-	if err != nil {
-		return nil, trace.Wrap(err)
-	}
-	return &resource{
-		reader: reader,
-		file:   f,
-	}, nil
+	const webAssetsMissingError = "the teleport binary was built without web assets, try building with `make release`"
+	return nil, trace.NotFound(webAssetsMissingError)
```
Comment (awly) [inline lib/web/static.go:30]:
```suggestion
	return nil, trace.NotFound(webAssetsMissingError)
```

File: lib/web/static_embed.go [hunk 4]
```diff
@@ -0,0 +1,34 @@
+// +build webassets_embed
+
+/*
+Copyright 2015 Gravitational, Inc.
```
Comment (awly) [inline lib/web/static_embed.go:4]:
```suggestion
Copyright 2021 Gravitational, Inc.
```

**Review by awly — APPROVED** [review]

**Comment by webvictim** [comment]
[BoringCrypto release list](https://go.googlesource.com/go/+/dev.boringcrypto/misc/boring/RELEASES)

File: build.assets/Dockerfile (hunk 2 above)
Comment (a-palchikov) [reply build.assets/Dockerfile:100]: As the required tag drifts away from tip, this will no longer work with `depth=1`.

**Comment by a-palchikov** [comment]
> [BoringCrypto release list](https://go.googlesource.com/go/+/dev.boringcrypto/misc/boring/RELEASES)

It's being all merged into https://github.com/golang/go/tree/dev.boringcrypto.go1.16 - is there anything specific about the release list that I should account for?

**Comment by webvictim** [comment]
> > [BoringCrypto release list](https://go.googlesource.com/go/+/dev.boringcrypto/misc/boring/RELEASES)
> 
> It's being all merged into https://github.com/golang/go/tree/dev.boringcrypto.go1.16 - is there anything specific about the release list that I should account for?

No, I was just posting it to make it easier to keep an eye on when 1.16.2  comes out (as the file will be updated to show that)

**Review by webvictim — CHANGES_REQUESTED** [review]

File: .drone.yml [hunk 5]
```diff
@@ -4,7 +4,7 @@ type: kubernetes
 name: test
 
 environment:
-  RUNTIME: go1.15.5
```
[the remark this answers is not in the document]
Comment (a-palchikov) [reply .drone.yml:7]: Yes, and I thought I did. Will use the dronegen instead - thanks.

**Review by webvictim — APPROVED** [review]

**Review by russjones — APPROVED** [review]
Bot.

**Comment by russjones** [comment]
@a-palchikov One change you're going to have to make is the dev.boringcrypto package name has been changed from a b5 suffix to b7 here:

https://github.com/gravitational/teleport/blob/master/build.assets/Dockerfile-fips#L37
https://go-boringcrypto.storage.googleapis.com/go1.16.2b7.linux-amd64.tar.gz

**Comment by a-palchikov** [comment]
> @a-palchikov One change you're going to have to make is the dev.boringcrypto package name has been changed from a b5 suffix to b7 here:
> 
> https://github.com/gravitational/teleport/blob/master/build.assets/Dockerfile-fips#L37
> https://go-boringcrypto.storage.googleapis.com/go1.16.2b7.linux-amd64.tar.gz

Ok, but this should change be made in context of linux FIPS builds - I can create a PR for this.

**Pull request merged**